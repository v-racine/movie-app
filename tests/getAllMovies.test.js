const { Config } = require('../src/config/config');

Config.Get(process.env);

const AppFactory = require('../src/app');
const request = require('supertest');

describe('get all movies', () => {
  const mockPgClient = {};

  const app = AppFactory({
    pgClient: mockPgClient,
  });

  describe('when: a user wants to view all movies with NO query strings', () => {
    let rsp;

    let movies = {
      rows: [
        {
          id: 1,
          movie_title: 'Nosferatu',
          movie_year: 2024,
          run_time: 132,
        },
        {
          id: 2,
          movie_title: 'Point Break',
          movie_year: 1991,
          run_time: 122,
        },
      ],
    };

    beforeEach(async () => {
      mockPgClient.dbQuery = jest.fn().mockImplementation(() => {
        return new Promise((resolve) => {
          resolve(movies);
        });
      });

      rsp = await request(app).get('/movies').send();
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    test('then: we return all the movie titles', async () => {
      expect(mockPgClient.dbQuery).toHaveBeenCalledWith('SELECT * FROM movies ORDER BY movie_year');

      const text = rsp.text;
      expect(text).toMatchSnapshot();
      // expect(text).toEqual(JSON.stringify(movies.rows));

      const status = rsp.status;
      expect(status).toBe(200);
    });
  });

  describe('when: a user want to view all movies with a specific release year', () => {
    let rsp;

    let movies = {
      rows: [
        { id: 1, title: 'There Will Be Blood', year: 2007, runTime: 158 },
        { id: 2, title: 'No Country for Old Men', year: 2007, runTime: 122 },
      ],
    };

    beforeEach(async () => {
      mockPgClient.dbQuery = jest.fn().mockImplementation(() => {
        return new Promise((resolve) => {
          resolve(movies);
        });
      });

      rsp = await request(app).get('/movies?year=2007').send();
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    test('then: we return all the movie titles with the specified release year', () => {
      expect(mockPgClient.dbQuery).toHaveBeenCalledWith(
        'SELECT * FROM movies WHERE movie_title = $1 OR movie_year = $2 OR run_time = $3 ORDER BY movie_year',
        undefined,
        '2007',
        undefined,
      );

      const text = rsp.text;
      expect(text).toMatchSnapshot();

      const status = rsp.status;
      expect(status).toBe(200);
    });
  });
});
