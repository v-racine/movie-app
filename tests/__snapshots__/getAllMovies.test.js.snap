// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get all movies when: a user wants to view all movies with NO query strings then: we return all the movie titles 1`] = `
<!DOCTYPE html>
<html lang="en-US">
  <head>
    <title>
      The Ultimate Movie App
    </title>
    <meta charset="UTF-8">
    <link rel="stylesheet"
          href="/stylesheets/application.css"
    >
  </head>
  <body>
    <header>
      <h1>
        The Ultimate Movie App
      </h1>
    </header>
    <main>
      <form action="/movies/create"
            method="get"
      >
        <button>
          Add a Movie
        </button>
      </form>
      <ul>
      </ul>
      <p>
      </p>
      <a href="/movies/1">
        Nosferatu
      </a>
      <p>
      </p>
      <a href="/movies/2">
        Point Break
      </a>
    </main>
  </body>
</html>
`;

exports[`get all movies when: a user wants to view all movies with a specific release year AND run time then: we return all the movie titles with the specified release year AND run time 1`] = `
<!DOCTYPE html>
<html lang="en-US">
  <head>
    <title>
      The Ultimate Movie App
    </title>
    <meta charset="UTF-8">
    <link rel="stylesheet"
          href="/stylesheets/application.css"
    >
  </head>
  <body>
    <header>
      <h1>
        The Ultimate Movie App
      </h1>
    </header>
    <main>
      <form action="/movies/create"
            method="get"
      >
        <button>
          Add a Movie
        </button>
      </form>
      <ul>
      </ul>
      <p>
      </p>
      <a href="/movies/1">
      </a>
      <p>
      </p>
      <a href="/movies/2">
      </a>
      <p>
      </p>
      <a href="/movies/3">
      </a>
    </main>
  </body>
</html>
`;

exports[`get all movies when: a user wants to view all movies with a specific release year BUT the user does NOT enter a valid year in the query string then: we return a flash error message that says 'Year must be a 4-digit number' 1`] = `
<!DOCTYPE html>
<html lang="en-US">
  <head>
    <title>
      The Ultimate Movie App
    </title>
    <meta charset="UTF-8">
    <link rel="stylesheet"
          href="/stylesheets/application.css"
    >
  </head>
  <body>
    <header>
      <h1>
        The Ultimate Movie App
      </h1>
    </header>
    <main>
      <ul>
        <li class="flash error">
          Year must be a 4-digit number
        </li>
      </ul>
      <p>
        Tell us about your favourite movies!
        <form action="/signUp"
              method="get"
        >
          <button>
            Sign Up
          </button>
        </form>
        <form action="/signIn"
              method="get"
        >
          <button>
            Sign In
          </button>
        </form>
        <form action="/movies"
              method="get"
        >
          <button>
            See Movies
          </button>
        </form>
        <form action="/reviews"
              method="get"
        >
          <button>
            See Reviews
          </button>
        </form>
      </p>
      <img src="/images/Children_of_paradise.jpg"
           alt="Children of Paradise"
           title="Children of Paradise"
      >
    </main>
  </body>
</html>
`;

exports[`get all movies when: a user wants to view all movies with a specific release year then: we return all the movie titles with the specified release year 1`] = `
<!DOCTYPE html>
<html lang="en-US">
  <head>
    <title>
      The Ultimate Movie App
    </title>
    <meta charset="UTF-8">
    <link rel="stylesheet"
          href="/stylesheets/application.css"
    >
  </head>
  <body>
    <header>
      <h1>
        The Ultimate Movie App
      </h1>
    </header>
    <main>
      <form action="/movies/create"
            method="get"
      >
        <button>
          Add a Movie
        </button>
      </form>
      <ul>
      </ul>
      <p>
      </p>
      <a href="/movies/1">
      </a>
      <p>
      </p>
      <a href="/movies/2">
      </a>
    </main>
  </body>
</html>
`;
